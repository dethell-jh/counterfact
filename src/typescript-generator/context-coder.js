import nodePath from "node:path";

import { Coder } from "./coder.js";

export class ContextCoder extends Coder {
  isRoot() {
    return this.requirement.url.split("/").at(-2).split("~1").length === 2;
  }

  names() {
    return super.names("Context");
  }

  write() {
    if (this.isRoot()) {
      return "{}";
    }

    return { raw: 'export { default } from "../$context.js"' };
  }

  modulePath() {
    const pathString = this.requirement.url
      .split("/")
      .at(-2)
      .replaceAll("~1", "/");

    return `${nodePath.join(
      "paths",
      nodePath.dirname(pathString)
    )}/$context.ts`;
  }
}
