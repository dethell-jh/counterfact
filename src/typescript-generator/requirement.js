export class Requirement {
  constructor(data, url = "", specification = undefined) {
    if (data === undefined) {
      throw new Error("cannot create requirement without data");
    }

    this.data = data;
    this.url = url;
    this.specification = specification;
  }

  get isReference() {
    return this.data?.$ref !== undefined;
  }

  reference() {
    return this.specification.requirementAt(this.data.$ref, this.url);
  }

  has(item) {
    return item in this.data;
  }

  get(item) {
    if (!this.has(item)) {
      return undefined;
    }

    return new Requirement(
      this.data[item],
      `${this.url}/${item}`,
      this.specification
    );
  }

  select(path, data = this.data, basePath = "") {
    const [head, ...tail] = path.split("/");

    const branch = data[this.unescapeJsonPointer(head)];

    if (!branch) {
      return undefined;
    }

    if (tail.length === 0) {
      return new Requirement(
        branch,
        `${this.url}/${basePath}${head}`,
        this.specification
      );
    }

    return this.select(tail.join("/"), branch, `${basePath}${head}/`);
  }

  toJSON() {
    return {
      specification: this.specification,
      url: this.url,
      data: this.data,
    };
  }

  forEach(callback) {
    Object.keys(this.data).forEach((key) => {
      callback([key, this.select(this.escapeJsonPointer(key))]);
    });
  }

  map(callback) {
    const result = [];

    // eslint-disable-next-line array-callback-return
    this.forEach(([key, value]) => result.push(callback(value, key)));

    return result;
  }

  flatMap(callback) {
    const result = [];

    // eslint-disable-next-line array-callback-return
    this.forEach(([key, value]) => result.push(callback(value, key)));

    return result.flat();
  }

  escapeJsonPointer(string) {
    return string.replaceAll("~", "~0").replaceAll("/", "~1");
  }

  unescapeJsonPointer(pointer) {
    return pointer.replaceAll("~1", "/").replaceAll("~0", "~");
  }
}
