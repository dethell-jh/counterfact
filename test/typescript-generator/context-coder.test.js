import prettier from "prettier";

import { ContextCoder } from "../../src/typescript-generator/context-coder.js";
import { Requirement } from "../../src/typescript-generator/requirement.js";

function format(code) {
  return prettier.format(code, { parser: "typescript" });
}

const dummyScript = {
  import() {
    return "schema";
  },

  importDefault() {
    return "default";
  },

  importExternalType() {
    return "ExternalType";
  },

  importType() {
    return "Type";
  },

  path: ".",

  repository: {
    get() {
      return {
        exportDefault() {
          return "noop";
        },
      };
    },
  },
};

describe("a ContextCoder", () => {
  it("generates a list of potential names", () => {
    const coder = new ContextCoder({ url: "#/components/schemas/Person" });

    const [one, two, three] = coder.names();

    expect([one, two, three]).toStrictEqual([
      "Context",
      "Context2",
      "Context3",
    ]);
  });

  it("when asked to delegate an inline requirement, returns itself", async () => {
    const coder = new ContextCoder(new Requirement({ type: "string" }));

    await expect(coder.delegate()).resolves.toBe(coder);
  });

  it("has a static ID", () => {
    const coder = new ContextCoder({ url: "#/paths/hello/get" });

    expect(coder.id).toBe("ContextCoder");
  });

  it("returns the context object if this is the root directory", async () => {
    const coder = new ContextCoder(new Requirement({}, "#/paths/hello"));

    const rootContextScript = {
      ...dummyScript,
      path: "paths/$.context.ts",
    };

    await expect(format(coder.write(rootContextScript))).resolves.toBe(
      await format("new Context()"),
    );
  });

  it("writes out the Context class", () => {
    const coder = new ContextCoder(new Requirement({}, "#/paths/hello"));

    expect(coder.beforeExport(dummyScript)).toContain("class Context {");
  });
});
