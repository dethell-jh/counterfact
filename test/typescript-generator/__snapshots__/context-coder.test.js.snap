// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a ContextCoder extends the Context from the parent directory 1`] = `
"export { default } from \\"../$.context.mjs\\";
"
`;

exports[`a ContextCoder finds the request method in the URL 1`] = `
"export { default } from \\"../$.context.mjs\\";
"
`;
